src/forklift/core.py

```diff
@@ -142,50 +144,58 @@ def truncate_where_clause(txt):
-            #: strip off duplicated primary key added during hashing since it's no longer necessary
-            adds_clause = changes.get_adds_where_clause(crate.source_primary_key, crate.source_primary_key_type, reproject_temp_suffix)
-            log.debug('adds where clause: %s', truncate_where_clause(adds_clause))
+            while changes.has_adds():
+                #: strip off duplicated primary key added during hashing since it's no longer necessary
+                adds_clause = changes.get_adds_where_clause(add_clause_key, add_clause_key_type, reproject_temp_suffix)
+                log.debug('adds where clause: %s', truncate_where_clause(adds_clause))
 
-            if not crate.is_table():
-                shape_field_index = -2
-                changes.fields[shape_field_index] = changes.fields[shape_field_index].rstrip('WKT')
+                if not crate.is_table():
+                    shape_field_index = -2
+                    changes.fields[shape_field_index] = changes.fields[shape_field_index].rstrip('WKT')
 
-            fields = changes.fields[:-1]
+                fields = changes.fields[:-1]
 
-            #: cache this so we don't have to call it for every record
-            is_table = crate.is_table()
+                #: cache this so we don't have to call it for every record
+                is_table = crate.is_table()
 
-            with arcpy.da.SearchCursor(changes.table, changes.fields, where_clause=adds_clause) as add_cursor,\
-                    arcpy.da.InsertCursor(crate.destination, fields) as cursor, \
-                    arcpy.da.InsertCursor(hash_table, [hash_id_field, hash_att_field, hash_geom_field]) as hash_cursor:
-                for row in add_cursor:
-                    primary_key = row[-1]
+                with arcpy.da.SearchCursor(changes.table, changes.fields, where_clause=adds_clause) as add_cursor,\
+                        arcpy.da.InsertCursor(crate.destination, fields) as cursor, \
+                        arcpy.da.InsertCursor(hash_table, [hash_id_field, hash_att_field, hash_geom_field]) as hash_cursor:
+                    for row in add_cursor:
+                        primary_key = row[-1]
 
-                    #: skip null geometries
-                    if not is_table and row[shape_field_index] is None:
-                        continue
+                        #: skip null geometries
+                        if not is_table and row[shape_field_index] is None:
+                            continue
 
-                    dest_id = cursor.insertRow(row[:-1])
-                    if crate.source_describe.hasOID and crate.source_describe.OIDFieldName == crate.source_primary_key:
-                        hash_key = dest_id
-                    else:
-                        hash_key = primary_key
+                        dest_id = cursor.insertRow(row[:-1])
+                        if crate.source_describe.hasOID and crate.source_describe.OIDFieldName == crate.source_primary_key:
```

Comment:
I don't think this check is needed either. The hash key should always be the OBJECTID of destination.